sycl/source/detail/platform_impl.cpp

```diff
@@ -233,14 +230,121 @@ std::shared_ptr<device_impl> platform_impl::getOrMakeDeviceImpl(
   return Result;
 }
 
+static bool supportsAffinityDomain(const device &dev,
+                                   info::partition_property partitionProp,
+                                   info::partition_affinity_domain domain) {
+  if (partitionProp != info::partition_property::partition_by_affinity_domain) {
+    return true;
+  }
+  auto supported = dev.get_info<info::device::partition_affinity_domains>();
+  for (info::partition_affinity_domain dom : supported) {
+    if (dom == domain) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool supportsPartitionProperty(const device &dev,
+                                      info::partition_property partitionProp) {
+  auto supported = dev.get_info<info::device::partition_properties>();
+  for (info::partition_property prop : supported) {
+    if (prop == partitionProp) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static std::vector<device>
+amendDeviceAndSubDevices(backend PlatformBackend,
+                         std::vector<device> &DeviceList,
+                         ods_target_list *OdsTargetList) {
+  constexpr info::partition_property partitionProperty =
+      info::partition_property::partition_by_affinity_domain;
+  constexpr info::partition_affinity_domain affinityDomain =
+      info::partition_affinity_domain::next_partitionable;
+
+  std::vector<device> FinalResult;
+
+  for (unsigned i = 0; i < DeviceList.size(); i++) {
+    // device has already been screened. The question is whether it should be a
+    // top level device and/or is expected to add its sub-devices to the list.
+    device &dev = DeviceList[i];
+    bool deviceAdded = false;
+    for (ods_target target : OdsTargetList->get()) {
+      backend TargetBackend = target.Backend;
+      if (PlatformBackend == TargetBackend || TargetBackend == backend::all) {
+        bool deviceMatch = target.HasDeviceWildCard; // opencl:*
+        if (target.HasDeviceType) {                  // opencl:gpu
+          deviceMatch = ((target.DeviceType == info::device_type::all) ||
+                         (dev.get_info<info::device::device_type>() ==
+                          target.DeviceType));
+
+        } else if (target.HasDeviceNum) { // opencl:0
+          deviceMatch = (target.DeviceNum == i);
+        }
+
+        if (deviceMatch) {
```

Comment:
Could this be an early continue instead? I.e. something like:
```c++
if (!deviceMatch)
  continue;
```

Reply:
Yes, but it might be personal preference. In this case, because the overall function is long, I prefer to keep the nesting. That way, if you someday someone is deep in the code wondering why something isn't working, they can look up the nesting and see every if() conditional that applies to them, giving them the correct mental model.   But if we do premature exit, then that is outside the nesting, so they have to read the nested clauses AND this premature exit that is outside and preceded them.  

So, I'd prefer to keep the nesting in this case.  However, as a compromise that will make both of us unhappy, I'd be willing to use a `goto` statement instead.

Reply:
I agree that the early exit does break the pathing somewhat, but only at the outer level. The benefit of having it is a reduction in nesting, making the whole tree less daunting... But the current version is better than a `goto`, so see this as a nit. 😅